source/common/ssl/ssl_socket.cc

```diff
@@ -402,39 +402,75 @@ std::string SslSocket::subjectLocalCertificate() const {
   return getSubjectFromCertificate(cert);
 }
 
+namespace {
+SslSocketFactoryStats generateStats(const std::string& prefix, Stats::Scope& store) {
+  return {
+      ALL_SSL_SOCKET_FACTORY_STATS(POOL_COUNTER_PREFIX(store, prefix + "_ssl_socket_factory."))};
+}
+} // namespace
+
 ClientSslSocketFactory::ClientSslSocketFactory(ClientContextConfigPtr config,
                                                Ssl::ContextManager& manager,
                                                Stats::Scope& stats_scope)
-    : manager_(manager), stats_scope_(stats_scope), config_(std::move(config)),
-      ssl_ctx_(manager_.createSslClientContext(stats_scope_, *config_)) {}
+    : manager_(manager), stats_scope_(stats_scope), stats_(generateStats("client", stats_scope)),
+      config_(std::move(config)),
+      ssl_ctx_(manager_.createSslClientContext(stats_scope_, *config_)) {
+  config_->setSecretUpdateCallback([this]() { onAddOrUpdateSecret(); });
+}
 
 Network::TransportSocketPtr ClientSslSocketFactory::createTransportSocket() const {
-  if (ssl_ctx_) {
-    return std::make_unique<Ssl::SslSocket>(ssl_ctx_, Ssl::InitialState::Client);
+  // onAddOrUpdateSecret() could be invoked in the middle of checking the existence of ssl_ctx and
+  // creating SslSocket using ssl_ctx. Capture ssl_ctx_ into a local variable so that we check and
+  // use the same ssl_ctx to create SslSocket.
+  auto ssl_ctx = ssl_ctx_;
+  if (ssl_ctx) {
+    return std::make_unique<Ssl::SslSocket>(std::move(ssl_ctx), Ssl::InitialState::Client);
   } else {
+    ENVOY_LOG(debug, "Create NotReadySslSocket");
+    stats_.upstream_connection_reset_by_sds_.inc();
```

Comment:
This stat name is confusing to me. Maybe rename to "upstream_context_secrets_not_ready" or something.

Reply:
Changed to upstream_context_secrets_not_ready